### PR TITLE
Feat/408716  in breach reason

### PR DESCRIPTION
**Pull request by whitewaterdesign**

### CODE REVIEW

File: app/api/ddi-index-api/dog-breaches.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * @property {string} short_name
+ */
+
+/**
+ * @return {Promise<BreachCategory[]>}
+ */
+const getBreachCategories = async () => {
+  const payload = await get(`${dogBreachesEndpoint}/categories`, options)
+
+  return payload.breachCategories.map(breachCategory => {
```
Comment (jbarnsley10) [inline app/api/ddi-index-api/dog-breaches.js:22]: I would personally avoid including functional programming here. It could be quite confusing if people haven't come across it before (like me :) ). Keep is consistent and simple in my view.

Comment (jbarnsley10) [reply app/api/ddi-index-api/dog-breaches.js:22]: Actually this isn't, it just looks like it based on the names you've used. Slightly confused - haha

File: app/routes/cdo/edit/change-status.js [hunk 2]
```diff
@@ -54,11 +72,16 @@ module.exports = [
...
-        payload: validatePayload,
+        payload: validateChangeStatusPayload,
         failAction: changeStatusPostFailAction
       },
       handler: async (request, h) => {
         const payload = request.payload
+        const backNav = addBackNavigation(request, false, true)
+
+        if (payload.newStatus === 'In breach') {
+          return h.redirect(`${routes.inBreach.get}/${payload.indexNumber}?src=${backNav?.srcHashValue}`)
```
Comment (jbarnsley10) [inline app/routes/cdo/edit/change-status.js:83]: I think this redirect may lose the back nav (or at least make it work wrongly) as you're creating a new srcHashParam. Worth testing this specific back nav case

File: app/routes/cdo/edit/change-status.js [hunk 3]
```diff
@@ -54,11 +72,16 @@ module.exports = [
     options: {
       auth: { scope: anyLoggedInUser },
       validate: {
-        payload: validatePayload,
+        payload: validateChangeStatusPayload,
         failAction: changeStatusPostFailAction
       },
       handler: async (request, h) => {
         const payload = request.payload
+        const backNav = addBackNavigation(request, false, true)
```
Comment (jbarnsley10) [inline app/routes/cdo/edit/change-status.js:80]: Too many params on the call to addBackNav??

File: app/routes/cdo/edit/change-status.js [hunk 4]
```diff
@@ -73,5 +96,42 @@ module.exports = [
...
+    path: `${routes.inBreach.get}/{indexNumber}/{dummy?}`,
+    options: {
+      auth: { scope: anyLoggedInUser },
+      handler: async (request, h) => {
+        const cdo = await getCdo(request.params.indexNumber)
+
+        if (cdo == null) {
+          return h.response().code(404).takeover()
+        }
+        const backNav = addBackNavigation(request, false, true)
```
Comment (jbarnsley10) [inline app/routes/cdo/edit/change-status.js:111]: Too many params again?

File: app/routes/cdo/edit/change-status.js [hunk 5]
```diff
@@ -73,5 +96,42 @@ module.exports = [
...
+        payload: validateBreachReasonPayload,
+        failAction: breachReasonPostFailAction
+      },
+      auth: { scope: anyLoggedInUser },
+      handler: async (request, h) => {
+        const payload = request.payload
+
+        await setDogBreaches(payload, getUser(request))
+
+        return h.redirect(`${routes.changeStatusConfirmation.get}/${payload.indexNumber}`)
```
Comment (jbarnsley10) [inline app/routes/cdo/edit/change-status.js:133]: You might want to keep the srcHashParam for the link to view dog record (so it keeps the proper history) but also not important if it is lost here

File: app/models/builders/builders.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
+/**
```
Comment (jbarnsley10) [inline app/models/builders/builders.js]: Not sure why these are needed

File: app/models/builders/components.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
+/**
```
Comment (jbarnsley10) [inline app/models/builders/components.js:1]: Is this file needed?

File: app/models/cdo/edit/change-status.js [hunk 8]
```diff
@@ -5,6 +5,7 @@ function ViewModel (dog, backNav, errors) {
   this.model = {
     backLink: backNav.backLink,
     srcHashParam: backNav.srcHashParam,
```
Comment (jbarnsley10) [inline app/models/cdo/edit/change-status.js:7]: srcHashValue isn't needed (I don't think) as srcHashParam is the same but with '?src=' prefixed

File: app/routes/cdo/edit/change-status.js (hunk 2 above)
Comment (jbarnsley10) [inline app/routes/cdo/edit/change-status.js:83]: Use srcHashParam (instead of '?src=' + srcHashValue)

File: app/api/ddi-index-api/dog-breaches.js (hunk 1 above)
Comment (whitewaterdesign) [reply app/api/ddi-index-api/dog-breaches.js:22]: changed to
```const [firstLetter, ...restOfLetters] = breachCategory.label
const label = firstLetter.toUpperCase() + restOfLetters.join('')```

File: app/models/builders/builders.js (hunk 6 above)
Comment (whitewaterdesign) [reply app/models/builders/builders.js]: This file has been removed

File: app/models/builders/components.js (hunk 7 above)
Comment (whitewaterdesign) [reply app/models/builders/components.js:1]: Most definately, this declares the full interface for govukCheckbox - going to make life much easier when writing the model for components

File: app/models/cdo/edit/change-status.js (hunk 8 above)
Comment (whitewaterdesign) [reply app/models/cdo/edit/change-status.js:7]: was there for another reason (input hidden), but had removed that anyway

File: app/routes/cdo/edit/change-status.js (hunk 3 above)
Comment (whitewaterdesign) [reply app/routes/cdo/edit/change-status.js:80]: removed third arg from this and the block above as well.

File: app/routes/cdo/edit/change-status.js (hunk 2 above)
Comment (whitewaterdesign) [reply app/routes/cdo/edit/change-status.js:83]: yeah, we tested this one on the screenshare.
